Many updates to the tools

xexyl opened the pull request:
I apologise for the long commit log: I made a lot of commits and some
had elaborate details (some important and others less important) and
unfortunately the logs did not carry over. I've removed my name, the
date and the commit id as GitHub does to make it as short as possible
and I also removed other text.

*   MAX_WINNER_HANDLE -> MAX_HANDLE
*   num_authors -> author_count
*   txzchk: Minor rewording message for !exists(tar)
*   jinfochk and jauthchk: Add formed_timestamp check
*   jinfochk/jauthchk: Add formed_timestamp_usec test
*   Add min_timestamp check to jinfochk and jauthchk
*   Add seqcexit rule to Makefile + run make seqcexit

    Checks for seqcexit tool being installed and if it is it runs it on the
    source files. Now whenever source files are updated one can simply run
    make seqcexit and the job will be done.

*   Add timestamp_epoch check to jinfochk and jauthchk
*   Fix test_JSON/good/info.json.sorted

    Outdated mkiocccentry_version caused that test to fail.

*   Two bug fixes in jinfochk and jauthchk

    Array infinite loop when array is first: although arrays are not handled
    yet this commit should prevent the tools from entering an infinite loop
    when they are in the file ahead of everything else; it only happened if
    the array was first because when extracting the value if it (the array)
    wasn't first the strtok_r() call would be passed in NULL as expected:
    when the array is first it happily continued to call strtok_r(p, ":",
    &savefield); which means that p would once again be the same value until
    the program was terminated.

    Remove trailing whitespace from the value. This fixes the problem where
    a valid value of false (as a boolean) triggered an error since using
    strcmp() on it did not exactly match "false" (or "true"). Only happens
    at the end of the file i.e. when not with a ',' at the end (this because
    of the second argument passed into strtok_r() being ",\0" which does not
    have a newline - which I think was the problem).

*   Fix bug where asking for IOCCC winner handle fails

    To test the jauthchk (when I get to the arrays part) I have made the
    mkiocccentry-test.sh write five authors to the entry. But this in turn
    revealed a bug: "IOCCC winner handle of author #3 does not start with an
    alpha numeric character." I added the printing of the actual string
    (seems helpful to the user) and it most certainly did start with an

xexyl commented:
Decided to do it now so I can start out without any problems tomorrow. I realised also what I was in the middle of I cannot finish today so I decided I'd quickly sort this out. I'm going to reply to any replies you have made that I've not replied to and then I will be gone for the day.

xexyl commented:
Thanks for the merge. Hopefully the log went through okay this time.

lcn2 commented:
> Decided to do it now so I can start out without any problems tomorrow. I realised also what I was in the middle of I cannot finish today so I decided I'd quickly sort this out. I'm going to reply to any replies you have made that I've not replied to and then I will be gone for the day.

Let us know if this worked.

Also if we didn't reply to something critical, please re-ask.  But as far as we can tell for now, we have responded to everything that you have asked.

xexyl commented:
> > Decided to do it now so I can start out without any problems tomorrow. I realised also what I was in the middle of I cannot finish today so I decided I'd quickly sort this out. I'm going to reply to any replies you have made that I've not replied to and then I will be gone for the day.
> 
> Let us know if this worked.
> 
> Also if we didn't reply to something critical, please re-ask. But as far as we can tell for now, we have responded to everything that you have asked.

I was more concerned that I didn't reply to everything you replied to! I think I have but will look at it more tomorrow. Thanks and have a great afternoon!

xexyl commented:
And for some reason the log did not carry over again. What should be done you think?

xexyl commented:
For now I'll not worry about it but it is unfortunate since it makes it easier to identify what I did when (like adding tests and fixing bugs).

The humour I had is less of a problem of course.

xexyl commented:
FWI: There was a bug in the handling of author handles that meant that alphanumerical character in the first char was detected as not alphanumerical. The reason is you forgot to reset `j = 0` at the top so that depending on the previous author handle length it would see `\0`!

I discovered this when making the test script have five authors.

lcn2 commented:
> And for some reason the log did not carry over again. What should be done you think?

We did the GitHub 'SQUISH and MERGE' action, via the iOS app, as we have done before.

Is this a new problem, or have there been pull requests in the past that didn't work as well?

xexyl commented:
> > And for some reason the log did not carry over again. What should be done you think?
> 
> We did the GitHub 'SQUISH and MERGE' action, via the iOS app, as we have done before.
> 
> Is this a new problem, or have there been pull requests in the past that didn't work as well?

One of my earliest commits had this problem as well.

xexyl commented:
I mean earliest commits in the repo - not today.

xexyl commented:
As it happens though in this case it worked out better: I accidentally left out some text in the commit log so it would be incomplete. The irony is wonderful.

Anyway that's all for now. Hopefully the next pull request doesn't have this problem. Tomorrow I'll see if I can figure out the timestamp problem and hopefully add some of the other tests not yet implemented.

xexyl commented:
I'm shutting down and going to sleep now but I just saw you did another commit and I wonder if this is how the git log was supposed to be for you too:

```shell
    Improve extra data file check in mkiocccentry
```

Did you provide more than that? If you did there's clearly some kind of problem and something should be done though what it is I don't know. Anyway hope you have a great night! More from me tomorrow.

lcn2 commented:
> I'm shutting down and going to sleep now but I just saw you did another commit and I wonder if this is how the git log was supposed to be for you too:
> 
> ```
>     Improve extra data file check in mkiocccentry
> ```
> 
> Did you provide more than that? If you did there's clearly some kind of problem and something should be done though what it is I don't know. Anyway hope you have a great night! More from me tomorrow.

We did not provide that log entry.  That must be what GitHub's _squish and merge_ did.

xexyl commented:
> > I'm shutting down and going to sleep now but I just saw you did another commit and I wonder if this is how the git log was supposed to be for you too:
> > ```
> >     Improve extra data file check in mkiocccentry
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Did you provide more than that? If you did there's clearly some kind of problem and something should be done though what it is I don't know. Anyway hope you have a great night! More from me tomorrow.
> 
> We did not provide that log entry. That must be what GitHub's _squish and merge_ did.
> 
> Not sure how to fix this or prevent useful data from being lost. Again this may be because we used the iOS GitHub app.
> 
> It might be worth it for us to explicitly try a merge using just the GitHub web site via a browser and see if that works.

It seems that that might be the problem. Check this out:

```
Squash and merge your pull request commits

When you select the Squash and merge option on a pull request on GitHub.com, the pull request's commits are squashed into a single commit. Instead of seeing all of a contributor's individual commits from a topic branch, the commits are combined into one commit and merged into the default branch. Pull requests with squashed commits are merged using the fast-forward option.
```

There's more here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges

There's also a table of what is included though it seems like it didn't happen for my commit. Maybe the link provides some information that will help prevent this problem in the future? In the meantime do you have a recommendation I should do for this? Because the commits from yesterday do have some important information. I'm going to once again delete the fork and re-fork and try again but this could be a problem for the future.

Btw did you notice I added a `make seqcexit` so now to sequence exit codes all you have to do is run that. It looks like:

```makefile
SEQCEXIT = seqcexit
# ...
seqcexit:
        @HAVE_SEQCEXIT=`command -v seqcexit`; if [[ -z "$$HAVE_SEQCEXIT" ]]; then \
                echo 'If you have not bothered to install the seqcexit tool, then' 1>&2; \
                echo 'you may not run this rule.'; 1>&2; \
                exit 1; \
        fi
        ${SEQCEXIT} ${SRCFILES}
```

The `SRCFILES` I added when I added `make depend` and it looks like:

```makefile
OBJFILES = dbg.o util.o mkiocccentry.o iocccsize.o fnamchk.o txzchk.o jauthchk.o jinfochk.o \
        json.o jstrencode.o jstrdecode.o rule_count.o
SRCFILES = $(patsubst %.o,%.c,$(OBJFILES))
```

So all one has to do is update the `OBJFILES` variable e.g. when adding a new source file and the `SRCFILES` will be automatically generated. But if you did not notice this change then that's a problem with the way the git log is not showing up.

lcn2 commented:
> > > I'm shutting down and going to sleep now but I just saw you did another commit and I wonder if this is how the git log was supposed to be for you too:
> > > ```
> > >     Improve extra data file check in mkiocccentry
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Did you provide more than that? If you did there's clearly some kind of problem and something should be done though what it is I don't know. Anyway hope you have a great night! More from me tomorrow.
> > 
> > 
> > We did not provide that log entry. That must be what GitHub's _squish and merge_ did.
> > Not sure how to fix this or prevent useful data from being lost. Again this may be because we used the iOS GitHub app.
> > It might be worth it for us to explicitly try a merge using just the GitHub web site via a browser and see if that works.
> 
> It seems that that might be the problem. Check this out:
> 
> ```
> Squash and merge your pull request commits
> 
> When you select the Squash and merge option on a pull request on GitHub.com, the pull request's commits are squashed into a single commit. Instead of seeing all of a contributor's individual commits from a topic branch, the commits are combined into one commit and merged into the default branch. Pull requests with squashed commits are merged using the fast-forward option.
> ```
> 
> There's more here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges

We will look at this link when we wake up after a very long marathon session.

Looks like we might have caused this problem by our default choices.  Sorry!

> There's also a table of what is included though it seems like it didn't happen for my commit. Maybe the link provides some information that will help prevent this problem in the future? In the meantime do you have a recommendation I should do for this? Because the commits from yesterday do have some important information. I'm going to once again delete the fork and re-fork and try again but this could be a problem for the future.

Hopefully when we wake up and review the link, we can select the proper option to merge a pull request.

We recall there are several options, perhaps one of the options other than _squish and merge_ is what we need to do with your next pull request?

lcn2 commented:
> Btw did you notice I added a `make seqcexit` so now to sequence exit codes all you have to do is run that. It looks like:

Yes, we used that rule in commit cc1d71017fe8f1d7671efa37e6eec8c976d811a7, thanks @xexyl.

xexyl commented:
> > > > I'm shutting down and going to sleep now but I just saw you did another commit and I wonder if this is how the git log was supposed to be for you too:
> > > > ```
> > > >     Improve extra data file check in mkiocccentry
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Did you provide more than that? If you did there's clearly some kind of problem and something should be done though what it is I don't know. Anyway hope you have a great night! More from me tomorrow.
> > > 
> > > 
> > > We did not provide that log entry. That must be what GitHub's _squish and merge_ did.
> > > Not sure how to fix this or prevent useful data from being lost. Again this may be because we used the iOS GitHub app.
> > > It might be worth it for us to explicitly try a merge using just the GitHub web site via a browser and see if that works.
> > 
> > 
> > It seems that that might be the problem. Check this out:
> > ```
> > Squash and merge your pull request commits
> > 
> > When you select the Squash and merge option on a pull request on GitHub.com, the pull request's commits are squashed into a single commit. Instead of seeing all of a contributor's individual commits from a topic branch, the commits are combined into one commit and merged into the default branch. Pull requests with squashed commits are merged using the fast-forward option.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There's more here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges
> 
> We will look at this link when we wake up after a very long marathon session.
> 
> Looks like we might have caused this problem by our default choices. Sorry!

Can happen to anyone! :)


> 
> > There's also a table of what is included though it seems like it didn't happen for my commit. Maybe the link provides some information that will help prevent this problem in the future? In the meantime do you have a recommendation I should do for this? Because the commits from yesterday do have some important information. I'm going to once again delete the fork and re-fork and try again but this could be a problem for the future.
> 
> Hopefully when we wake up and review the link, we can select the proper option to merge a pull request.

Hopefully you can.

> 
> We recall there are several options, perhaps one of the options other than _squish and merge_ is what we need to do with your next pull request?

I wouldn't know: I'm not experienced with GitHub. But I guess it's possible. At the same time it might just do the same thing. I believe the earlier commit actually as only one commit that the log did not carry over for. However that was quite a while ago and I've changed a lot (and you have changed lots too) so maybe not.

Have a good rest or sleep or whatever you need! In the meantime I'm fixing the changes I made to work with what you did today. Hopefully I can get it sorted out soon. I do have to go do some morning stuff and I'll be taking a break at the hour (play a word game with my mum - we're very close). 

Speaking of my mum one of the reasons I'm unable to work as much for the time being (or as long) is when the Olympics are on we watch them together (though they're recorded and it'll go past the Olympics). That's part of why I'm not that upset that I'm waking up super early - it allows me more time to work on the code!

Hoping to have a pull request soon. I am not sure if I should once again delete my fork and set it p anew so that I can get everything right. I might do that after getting the files fixed (move the directory, delete the fork, re-fork, copy the changed files back).

xexyl commented:
> > Btw did you notice I added a `make seqcexit` so now to sequence exit codes all you have to do is run that. It looks like:
> 
> Yes, we used that rule in commit [cc1d710](https://github.com/ioccc-src/mkiocccentry/commit/cc1d71017fe8f1d7671efa37e6eec8c976d811a7), thanks @xexyl.

Welcome. Makes it much easier! I actually have a request for you but I'll write about that later on - it I think will require a change to that tool as well. More details later on today!

xexyl commented:
Btw in case it helps: this is the first time it happened:

```git
commit fa056deacbad0190cb02c00390f366a175a52932
Author: Cody Boone Ferguson <53008573+xexyl@users.noreply.github.com>
Date:   Fri Jan 28 12:21:24 2022 -0800

    Minor answers file fixes (#36)
```

Unfortunately I have no idea what I had actually done in that commit but I'm certain that I wrote more than that - I always do when I make a fix. Hopefully that helps - somehow.

lcn2 commented:
> It seems that that might be the problem. Check this out:
> 
> ```
> Squash and merge your pull request commits
> 
> When you select the Squash and merge option on a pull request on GitHub.com, the pull request's commits are squashed into a single commit. Instead of seeing all of a contributor's individual commits from a topic branch, the commits are combined into one commit and merged into the default branch. Pull requests with squashed commits are merged using the fast-forward option.
> ```
> 
> There's more here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges

According to [Squashing and merging a long-running branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squashing-and-merging-a-long-running-branch):

```
If you plan to continue work on the [head branch](https://docs.github.com/en/github/getting-started-with-github/github-glossary#head-branch) of a pull request after the pull request is merged, we recommend you don't squash and merge the pull request.
```

OK, so we don't want to do the _ Squash and merge_ process ...

xexyl commented:
> > It seems that that might be the problem. Check this out:
> > ```
> > Squash and merge your pull request commits
> > 
> > When you select the Squash and merge option on a pull request on GitHub.com, the pull request's commits are squashed into a single commit. Instead of seeing all of a contributor's individual commits from a topic branch, the commits are combined into one commit and merged into the default branch. Pull requests with squashed commits are merged using the fast-forward option.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There's more here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges
> 
> According to [Squashing and merging a long-running branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squashing-and-merging-a-long-running-branch):
> 
> ```
> If you plan to continue work on the [head branch](https://docs.github.com/en/github/getting-started-with-github/github-glossary#head-branch) of a pull request after the pull request is merged, we recommend you don't squash and merge the pull request.
> ```
> 
> OK, so we don't want to do the _ Squash and merge_ process ...

Hopefully whatever you end up doing works out okay then. I'm about to head off for the day - at least the computer. I started the string to int conversion functions but that's where I left off. Will do another pull request tomorrow with more checks added.

lcn2 commented:
> OK, so we don't want to do the _ Squash and merge_ process ...

Continuing to _think out loud_ ...

So when we are presented with a pull request, we see these options:

```
Don't worry, we won't select this option:
```

<img width="312" alt="pull-request-options" src="https://user-images.githubusercontent.com/10429707/154866535-29a495d7-d94c-4958-a31e-b829ac96f271.png">

We are not sure why _Squish and merge_ is the default.  But given [our previous remark](https://github.com/ioccc-src/mkiocccentry/pull/84#issuecomment-1046330822), we know we DO NOT want to do that ...

xexyl commented:
I'm not sure either. Nor am I sure what would be the best option. I am sure that you'll figure it out though.

lcn2 commented:
> I'm not sure either. Nor am I sure what would be the best option. I am sure that you'll figure it out though.

Our guess is that we want _Create a merge commit_.  What do you think @xexyl ?

xexyl commented:
> > I'm not sure either. Nor am I sure what would be the best option. I am sure that you'll figure it out though.
> 
> Our guess is that we want _Create a merge commit_. What do you think @xexyl ?

Looking at it again that's my guess as well. Maybe documentation is available that would answer it?

xexyl commented:
But I should also point out that I've never done a git rebase so I'm not familiar with the specifics of that. I probably should read up on that one of these days. If you want to delay it until later it's okay by me - like I said I'll be off in a few minutes.

lcn2 commented:
> > > I'm not sure either. Nor am I sure what would be the best option. I am sure that you'll figure it out though.
> > 
> > 
> > Our guess is that we want _Create a merge commit_. What do you think @xexyl ?
> 
> Looking at it again that's my guess as well. Maybe documentation is available that would answer it?

I'm looking ...  How about I try it for the pull request #85 ?  You can see what happens when we do.

xexyl commented:
> > > > I'm not sure either. Nor am I sure what would be the best option. I am sure that you'll figure it out though.
> > > 
> > > 
> > > Our guess is that we want _Create a merge commit_. What do you think @xexyl ?
> > 
> > 
> > Looking at it again that's my guess as well. Maybe documentation is available that would answer it?
> 
> I'm looking ... How about I try it for the pull request #85 ? You can see what happens when we do.

You can do that then. My only thought is that we might not be sure it solves the problem either - because this issue didn't always happen.

lcn2 commented:
What do you think @xexyl:

<img width="860" alt="step" src="https://user-images.githubusercontent.com/10429707/154866998-04a7a17e-83a0-49d5-bdd7-06bd41765ce2.png">

And from our end we see:

<img width="964" alt="fork" src="https://user-images.githubusercontent.com/10429707/154867011-9fe4fb9c-eed7-4838-a21c-d9e79a4e34cb.png">

And after a `git fetch ...` and `git merge --ff-only ...` and `git rebase --rebase-merges ...`:

<img width="1060" alt="git-op" src="https://user-images.githubusercontent.com/10429707/154867127-4a41952e-e332-462e-99a0-e9e8d15d81b1.png">

We see that our location git repo shows (this is the `fork` app on macOS):

<img width="975" alt="fork-display" src="https://user-images.githubusercontent.com/10429707/154867173-bb99647c-3fe4-4226-af23-07f47290d3fa.png">

Looks OK at first glance.

xexyl commented:
I guess we'll have to wait and see if it solves the problem but it looks okay on my end too. I pasted the output of `git log` in the other thread (not sure which one now though).

xexyl commented:
Thanks again! I'll make another pull request tomorrow. Going now. Have a great day!
